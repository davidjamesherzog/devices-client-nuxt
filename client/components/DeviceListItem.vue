<template>
  <div class="box" :data-list-item-id="device.id">
    <article class="media">
      <div class="media-left">
        <div class="content has-text-centered">
          <b-icon
            :icon="type"
            size="is-large"
            type="is-primary"
            data-card-icon
          />
        </div>
      </div>
      <div class="media-content">
        <div class="content">
          <p>
            <strong :data-list-item-title="device.id">{{ device.name }}</strong>
            <small :data-list-item-os="device.id">{{ device.os }}</small>
            <br />
            <span :data-list-item-description="device.id">
              {{ device.description }}
            </span>
          </p>
        </div>
      </div>
    </article>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'nuxt-property-decorator';
import Device from '~/client/models/Device';

@Component({
  name: 'DeviceListItem'
})
export default class DeviceListItem extends Vue {
  @Prop({ type: Object })
  private device!: Device;

  get type(): String {
    let icon: String = '';
    switch (this.device.os) {
      case 'Mac':
      case 'Windows':
        icon = 'laptop';
        break;
      case 'Linux':
        icon = 'linux';
        break;
      case 'Android':
        icon = 'android';
        break;
      case 'iOS':
        icon = 'apple-ios';
        break;
      case 'Samsung':
      case 'Vizio':
      case 'FireTV':
      case 'Roku':
        icon = 'television';
        break;
      case 'Xbox':
        icon = 'microsoft-xbox';
        break;
      case 'Playstation':
        icon = 'sony-playstation';
        break;
      case 'Chromecast':
        icon = 'cast';
        break;
      case 'Chromebook':
        icon = 'chrome';
        break;
      default:
        icon = '';
        break;
    }
    return icon;
  }
}
</script>

<style lang="scss" scoped>
.box {
  margin-bottom: 0.5rem;
  cursor: pointer;
}
</style>
