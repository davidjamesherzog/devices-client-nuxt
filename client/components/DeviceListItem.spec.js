import { createLocalVue, shallowMount } from '@vue/test-utils';
import Buefy from 'buefy';
import DeviceListItem from '~/components/DeviceListItem.vue';

describe('DeviceListItem', () => {
  describe('type', () => {
    test('should return laptop (Mac)', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Mac'
      };
      expect(deviceListItem.type).toEqual('laptop');
    });

    test('should return laptop (Windows)', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Windows'
      };
      expect(deviceListItem.type).toEqual('laptop');
    });

    test('should return linux', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Linux'
      };
      expect(deviceListItem.type).toEqual('linux');
    });

    test('should return android', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Android'
      };
      expect(deviceListItem.type).toEqual('android');
    });

    test('should return apple-ios', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'iOS'
      };
      expect(deviceListItem.type).toEqual('apple-ios');
    });

    test('should return television (Samsung)', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Samsung'
      };
      expect(deviceListItem.type).toEqual('television');
    });

    test('should return television (Vizio)', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Vizio'
      };
      expect(deviceListItem.type).toEqual('television');
    });

    test('should return television (FireTV)', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'FireTV'
      };
      expect(deviceListItem.type).toEqual('television');
    });

    test('should return television (Roku)', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Roku'
      };
      expect(deviceListItem.type).toEqual('television');
    });

    test('should return microsoft-xbox', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Xbox'
      };
      expect(deviceListItem.type).toEqual('microsoft-xbox');
    });

    test('should return sony-playstation', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Playstation'
      };
      expect(deviceListItem.type).toEqual('sony-playstation');
    });

    test('should return cast', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Chromecast'
      };
      expect(deviceListItem.type).toEqual('cast');
    });

    test('should return chrome', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Chromebook'
      };
      expect(deviceListItem.type).toEqual('chrome');
    });

    test('should return blank', () => {
      const deviceListItem = new DeviceListItem();
      deviceListItem.device = {
        os: 'Other'
      };
      expect(deviceListItem.type).toEqual('');
    });
  });

  describe('template', () => {
    let localVue;

    beforeEach(() => {
      localVue = createLocalVue();
      localVue.use(Buefy);
    });

    test('should render device list item', () => {
      const wrapper = shallowMount(DeviceListItem, {
        localVue,
        propsData: {
          device: {
            id: 1,
            name: 'Macbook-Dave',
            description: 'Macbook laptop',
            os: 'Mac'
          }
        }
      });
      const id = wrapper.find('[data-list-item-id="1"]');
      expect(id).toBeDefined();
      const title = wrapper.find('[data-list-item-title="1"]');
      expect(title.text()).toEqual('Macbook-Dave');
      const description = wrapper.find('[data-list-item-description="1"]');
      expect(description.text()).toEqual('Macbook laptop');
      const os = wrapper.find('[data-list-item-os="1"]');
      expect(os.text()).toEqual('Mac');
      expect(wrapper).toMatchSnapshot();
    });
  });
});
