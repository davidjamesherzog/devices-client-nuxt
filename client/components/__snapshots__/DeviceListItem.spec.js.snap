// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceListItem template should render device list item 1`] = `
<div data-list-item-id="1" class="box">
  <article class="media">
    <div class="media-left">
      <div class="content has-text-centered">
        <b-icon-stub type="is-primary" icon="laptop" size="is-large" data-card-icon=""></b-icon-stub>
      </div>
    </div>
    <div class="media-content">
      <div class="content">
        <p><strong data-list-item-title="1">Macbook-Dave</strong> <small data-list-item-os="1">Mac</small> <br> <span data-list-item-description="1">
            Macbook laptop
          </span></p>
      </div>
    </div>
  </article>
</div>
`;
